File: src/Systems/C++/chain_of_responsibility.cpp
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

class RefundHandler {
public:
    virtual ~RefundHandler() = default;

    RefundHandler& setNext(RefundHandler& next) {
        next_ = &next;
        return next;
    }

    std::string handle(int amount, std::vector<std::string>& visited) const {
        visited.push_back(name());
        if (canHandle(amount)) {
            return name();
        }
        if (next_ == nullptr) {
            throw std::runtime_error("No handler accepted the request.");
        }
        return next_->handle(amount, visited);
    }

protected:
    virtual std::string name() const = 0;
    virtual bool canHandle(int amount) const = 0;

private:
    RefundHandler* next_ = nullptr;
};

class FaqHandler final : public RefundHandler {
protected:
    std::string name() const override { return "faq"; }
    bool canHandle(int amount) const override { return amount <= 50; }
};

class BillingHandler final : public RefundHandler {
protected:
    std::string name() const override { return "billing"; }
    bool canHandle(int amount) const override { return amount <= 500; }
};

class EscalationHandler final : public RefundHandler {
protected:
    std::string name() const override { return "escalation"; }
    bool canHandle(int) const override { return true; }
};

int main() {
    FaqHandler faq;
    BillingHandler billing;
    EscalationHandler escalation;
    faq.setNext(billing).setNext(escalation);

    std::vector<std::string> visited;
    const auto handled = faq.handle(250, visited);
    std::ostringstream path;
    for (std::size_t i = 0; i < visited.size(); ++i) {
        if (i > 0) path << '>';
        path << visited[i];
    }

    std::cout << "visited=" << path.str()
              << ";handled=" << handled
              << ";result=refund(250)\n";
}

File: src/Systems/Swift/chain_of_responsibility.swift
final class Handler {
    let name: String
    let limit: Int
    var next: Handler?

    init(_ name: String, _ limit: Int) {
        self.name = name
        self.limit = limit
    }

    @discardableResult
    func then(_ handler: Handler) -> Handler {
        next = handler
        return handler
    }

    func handle(_ amount: Int, visited: inout [String]) -> String {
        visited.append(name)
        if amount <= limit || next == nil {
            return name
        }
        return next!.handle(amount, visited: &visited)
    }
}

let faq = Handler("faq", 50)
let billing = Handler("billing", 500)
let escalation = Handler("escalation", Int.max)
faq.then(billing).then(escalation)

var visited: [String] = []
let handled = faq.handle(250, visited: &visited)
print("visited=\(visited.joined(separator: ">"));handled=\(handled);result=refund(250)")
